llvm/tools/llvm-xargs/llvm-xargs.cpp

```diff
@@ -0,0 +1,140 @@
+    "in-replace",
+    cl::desc("Replace R in command with names read from standard input; if R "
+             "is unspecified, assume {}"),
+    cl::init("{}"), cl::value_desc("R")};
+
+static cl::opt<std::string> OutReplace{
+    "out-replace",
+    cl::desc("Replace R in command with temporary filename for writing outputs "
+             "from executed command"),
+    cl::init(""), cl::value_desc("R")};
+
+// Emit list of produced files for better integration with other tools.
+static cl::opt<std::string> OutputFileList{
+    "out-file-list", cl::desc("Specify filename for list of outputs."),
+    cl::value_desc("filename"), cl::init("")};
+
+static void error(const Twine &Msg) {
+  errs() << "llvm-xargs: " << Msg << '\n';
+  exit(1);
+}
+
+static void error(std::error_code EC, const Twine &Prefix) {
+  if (EC)
+    error(Prefix + ": " + EC.message());
+}
+
+int main(int argc, char **argv) {
+
+  cl::ParseCommandLineOptions(
+      argc, argv,
+      "llvm-xargs: Execute specified command as many times as\n"
+      "necessary to use up the list of input items.\n"
+      "Usage:\n"
+      "llvm-xargs a.list 'echo {}'\n"
+      "NOTE: filenames containing blanks or newlines and commands containig\n"
+      "redirects are not supported by llvm-xargs yet.\n");
+
+  ExitOnError ExitOnErr("llvm-xargs: ");
+  std::unique_ptr<MemoryBuffer> MB =
+      ExitOnErr(errorOrToExpected(MemoryBuffer::getFileOrSTDIN(InputFileList)));
+  std::string FileList = "";
+
+  // Split input command into program and args.
+  std::string CurCommand = InputCommand;
+  StringRef ResCommand = CurCommand;
+  SmallVector<StringRef, 4> Args;
+  ResCommand.split(Args, /*Separator*/ " ", /*MaxSplit*/ -1,
+                   /*KeepEmpty*/ false);
+  if (Args.empty())
+    error("No command?");
+
+  // Find args to replace with filenames from input list.
+  int InReplaceArg = -1;
+  int OutReplaceArg = -1;
+  for (size_t i = 0; i < Args.size(); ++i) {
+    if (Args[i] == Replace)
+      InReplaceArg = i;
+    if (Args[i] == OutReplace)
+      OutReplaceArg = i;
+  }
+
+  // Emit an error if user requested replace output file in the command but
+  // replace string is not found.
+  if (!OutReplace.empty() && OutReplaceArg < 0)
+    error("Couldn't find replace string for output in the command");
+
+  // Make sure that specified program exists, emit an error if not.
+  std::string Prog =
+      ExitOnErr(errorOrToExpected(sys::findProgramByName(Args[0])));
+
+  std::error_code EC;
+  for (line_iterator LI(*MB); !LI.is_at_eof(); ++LI) {
+    std::string FileName = LI->str();
+    if (InReplaceArg > 0)
+      Args[InReplaceArg] = FileName;
+
+    SmallString<128> Path;
+    if (!OutReplace.empty()) {
+      // Create a temporary file for command result. Add file name to output
+      // file list if needed.
+      std::string TempFileNameBase = sys::path::stem(FileName);
+      EC = sys::fs::createTemporaryFile(TempFileNameBase, "", Path);
+      error(EC, "");
```

Comment:
With an empty prefix this message is going to look weird. How about `error(EC, "Could not create a temporary file")`?

Reply:
Done.